File: app/src/main/kotlin/com/youcoded/app/skills/McpReconciler.kt
package com.youcoded.app.skills

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * MCP Reconciler (decomposition v3, §9.3)
 *
 * Kotlin mirror of desktop/src/main/mcp-reconciler.ts. Scans each installed
 * plugin's mcp-manifest.json and reconciles ~/.claude.json `mcpServers`.
 *
 * Android always reports platform "linux" for filtering — Termux's bionic
 * environment is Linux-binary-compatible, so manifests declaring
 * `platform: "linux"` / `platforms: [..."linux"...]` or `"all"` are eligible.
 * macOS-only and Windows-only entries are skipped.
 */
class McpReconciler(private val homeDir: File) {

    data class Result(
        val added: Int,
        val skippedPlatform: Int,
        val skippedManual: Int,
        val manifestCount: Int,
    )

    private val claudeJson = File(homeDir, ".claude.json")

    // Android always counts as linux for MCP platform filtering
    private val currentPlatform = "linux"

    private fun expandTokens(s: String, pluginRoot: File): String = expandTokens(s, pluginRoot.absolutePath)

    private fun readManifest(pluginDir: File): Pair<JSONArray, File>? {
        val f = File(pluginDir, "mcp-manifest.json")
        if (!f.exists()) return null
        return try {
            val raw = f.readText()
            // Manifest can be either a raw array or an object with { servers: [...] }
            val entries = when {
                raw.trimStart().startsWith("[") -> JSONArray(raw)
                else -> JSONObject(raw).optJSONArray("servers") ?: JSONArray()
            }
            entries to pluginDir
        } catch (e: Exception) {
            Log.w(TAG, "Malformed manifest: ${f.absolutePath}", e)
            null
        }
    }

    private fun listManifests(): List<Pair<JSONArray, File>> =
        // Scan both roots — marketplace plugins can declare MCP servers too.
        ClaudeCodeRegistry.listInstalledPluginDirs(homeDir)
            .mapNotNull { readManifest(it) }

    private fun readClaudeJson(): JSONObject =
        try { if (claudeJson.exists()) JSONObject(claudeJson.readText()) else JSONObject() }
        catch (_: Exception) { JSONObject() }

    private fun writeClaudeJsonAtomic(data: JSONObject) {
        val tmp = File(claudeJson.parentFile, "${claudeJson.name}.${android.os.Process.myPid()}.tmp")
        tmp.writeText(data.toString(2))
        if (!tmp.renameTo(claudeJson)) {
            claudeJson.writeText(data.toString(2))
            tmp.delete()
        }
    }

    private fun buildServerConfig(entry: JSONObject, pluginRoot: File): JSONObject? =
        buildServerConfig(entry, pluginRoot.absolutePath, ::expandTokens)

    private fun buildServerConfig(
        entry: JSONObject,
        pluginRoot: String,
        expand: (String, String) -> String,
    ): JSONObject? {
        val type = entry.optString("type", "stdio")
        if (type == "http") {
            val url = entry.optString("url")
            if (url.isEmpty()) return null
            return JSONObject().put("type", "http").put("url", url)
        }
        // stdio — Android picks `command` (no command_windows variant applies)
        val cmd = entry.optString("command")
        if (cmd.isEmpty()) return null
        val config = JSONObject()
            .put("type", "stdio")
            .put("command", expand(cmd, pluginRoot))
        val args = entry.optJSONArray("args")
        if (args != null) {
            val expanded = JSONArray()
            for (i in 0 until args.length()) {
                expanded.put(expand(args.optString(i), pluginRoot))
            }
            config.put("args", expanded)
        }
        val env = entry.optJSONObject("env")
        if (env != null) config.put("env", env)
        return config
    }

    fun reconcile(): Result {
        val manifests = listManifests()
        val claude = readClaudeJson()
        val servers = claude.optJSONObject("mcpServers") ?: JSONObject().also {
            claude.put("mcpServers", it)
        }

        val r = applyManifestEntries(servers, manifests.map { (e, root) -> e to root.absolutePath })
        if (r.changed) writeClaudeJsonAtomic(claude)
        return Result(r.added, r.skippedPlatform, r.skippedManual, manifests.size)
    }

    data class ScanResult(
        val added: Int,
        val repaired: Int,
        val skippedPlatform: Int,
        val skippedManual: Int,
    ) {
        val changed: Boolean get() = added + repaired > 0
    }

    /**
     * Pure manifest-scan step (mirror of desktop `applyManifestEntries`,
     * exposed for tests). Mutates [servers]. Additive-only, except that an
     * untouched entry an older build wrote with a literal `${PACKAGE_DIR}`
     * is replaced — WHY: that entry never worked (nothing expanded the token),
     * and without the repair every existing install keeps it forever because
     * the scan never overwrites. Deep equality with the old output means an
     * entry the user edited in any way is still left alone.
     */
    internal fun applyManifestEntries(
        servers: JSONObject,
        manifests: List<Pair<JSONArray, String>>,
    ): ScanResult {
        var added = 0
        var repaired = 0
        var skippedPlatform = 0
        var skippedManual = 0
        for ((entries, pluginRoot) in manifests) {
            for (i in 0 until entries.length()) {
                val entry = entries.optJSONObject(i) ?: continue
                val name = entry.optString("name")
                if (name.isEmpty()) continue
                if (!platformMatches(entry, currentPlatform)) { skippedPlatform++; continue }
                if (!entry.optBoolean("auto", false)) { skippedManual++; continue }
                val config = buildServerConfig(entry, pluginRoot, ::expandTokens) ?: continue
                val existing = servers.optJSONObject(name)
                if (servers.has(name)) {
                    // Never overwrite user-configured entries (repair exception above).
                    val legacy = buildServerConfig(entry, pluginRoot, ::legacyExpand)
                    if (existing != null && legacy != null &&
                        legacy.toString().contains(UNEXPANDED_PACKAGE_DIR) &&
                        jsonEquals(existing, legacy)
                    ) {
                        servers.put(name, config)
                        repaired++
                    }
                    continue
                }
                servers.put(name, config)
                added++
            }
        }
        return ScanResult(added, repaired, skippedPlatform, skippedManual)
    }

    companion object {
        private const val TAG = "McpReconciler"

        private const val UNEXPANDED_PACKAGE_DIR = "\${PACKAGE_DIR}"

        /** WHY: `${PACKAGE_DIR}` is the token the marketplace publisher
         *  skill and two published manifests use; only `{{plugin_root}}` was
         *  expanded, so those servers were written with a literal placeholder
         *  and never started. Both now mean the plugin's install directory. */
        internal fun expandTokens(s: String, pluginRoot: String): String =
            s.replace("{{plugin_root}}", pluginRoot).replace(UNEXPANDED_PACKAGE_DIR, pluginRoot)

        /** What older builds wrote: `{{plugin_root}}` expanded, `${PACKAGE_DIR}` literal. */
        private fun legacyExpand(s: String, pluginRoot: String): String =
            s.replace("{{plugin_root}}", pluginRoot)

        // WHY: manifests use Node's names (darwin/win32) as well as the app's
        // (macos/windows); normalise both so neither is silently a non-match.
        private fun normalizePlatform(p: String): String = when (val v = p.trim().lowercase()) {
            "darwin", "mac", "osx" -> "macos"
            "win32", "win" -> "windows"
            else -> v
        }

        /** WHY: published manifests declare a `platforms` LIST, which the
         *  single `platform` field cannot express; reading only `platform`
         *  let those servers through everywhere. A non-empty list wins; an
         *  empty or missing one falls back to `platform`. */
        internal fun platformMatches(entry: JSONObject, current: String): Boolean {
            val list = entry.optJSONArray("platforms")
            if (list != null && list.length() > 0) {
                val names = (0 until list.length()).mapNotNull { list.opt(it) as? String }.map(::normalizePlatform)
                return "all" in names || current in names
            }
            val declared = entry.optString("platform")
            if (declared.isEmpty() || declared == "all") return true
            return normalizePlatform(declared) == current
        }

        // org.json has no structural equals; compare parsed values recursively
        // so key order in the user's file never matters.
        private fun jsonEquals(a: Any?, b: Any?): Boolean = when {
            a is JSONObject && b is JSONObject -> {
                // Android's org.json has no keySet(); keys() is the portable form.
                val ka = a.keys().asSequence().toSet()
                ka == b.keys().asSequence().toSet() && ka.all { jsonEquals(a.opt(it), b.opt(it)) }
            }
            a is JSONArray && b is JSONArray ->
                a.length() == b.length() && (0 until a.length()).all { jsonEquals(a.opt(it), b.opt(it)) }
            else -> a == b
        }
    }
}

File: desktop/src/main/mcp-reconciler.ts
import { app } from 'electron';
import fs from 'fs';
import path from 'path';
import os from 'os';
import { listInstalledPluginDirs } from './claude-code-registry';
import { NativeHome } from './native-home';
import { SecretsStore } from './providers/secrets-store';
import { McpRegistry, type ResolvedMcpServer } from './harness/mcp/mcp-registry';
import { log } from './logger';

/**
 * MCP Reconciler (decomposition v3 §9.3; native MCP phase 1, Task 7)
 *
 * Reconciles Claude Code's ~/.claude.json `mcpServers` section from TWO
 * sources:
 *
 *   1. Plugin manifests (~/.claude/plugins/ * /mcp-manifest.json) — the
 *      original, additive-only path. Only auto-registers `auto: true`
 *      entries, filtered by `platform` / `platforms`, expanding
 *      `{{plugin_root}}` and its alias `${PACKAGE_DIR}`. Never removes an
 *      existing entry and never overwrites one, except an untouched entry an
 *      older build wrote with a literal `${PACKAGE_DIR}` (see
 *      applyManifestEntries).
 *   2. The YouCoded MCP registry (~/.youcoded/mcp.json, via McpRegistry) —
 *      new in Task 7. See `projectToClaudeJson`'s header comment for the
 *      ownership rule that governs this source, INCLUDING the collision
 *      guard added in the fix pass: an id that already exists in mcpServers
 *      and that YouCoded did not previously own (hand-written by the user via
 *      `claude mcp add <id> ...`, or scanned from a plugin manifest) is never
 *      overwritten — it is skipped and reported, not silently replaced.
 *
 * Servers with `auto: false` (setup required, e.g., iMessages full-disk-access,
 * Todoist OAuth) are skipped — those need the user to act before they work,
 * and surfacing that belongs in the marketplace UI, not a silent reconciler.
 *
 * OPEN QUESTION (flagged in task-7-report.md, not resolved by this task): the
 * plan's Task 7 step 4 says the manifest scan "moves to feeding the registry".
 * That migration is deliberately NOT done here — see the report for why.
 */

const CLAUDE_JSON = path.join(os.homedir(), '.claude.json');

export interface McpManifestEntry {
  name: string;
  description?: string;
  platform?: 'macos' | 'windows' | 'linux' | 'all';
  /** WHY: published marketplace manifests (spotify-services,
   *  youcoded-messaging) — and the publisher skill that coaches authors —
   *  write a LIST of Node platform names (`["darwin", "win32"]`), which the
   *  single `platform` field cannot express. Reading only `platform` meant the
   *  filter silently let those servers through everywhere. Both spellings and
   *  both vocabularies are accepted; see platformMatches. */
  platforms?: string[];
  type?: 'stdio' | 'http';
  command?: string;
  command_windows?: string; // platform-specific override
  args?: string[];
  env?: Record<string, string>;
  url?: string;
  auto?: boolean;
  setup_note?: string;
}

interface ClaudeJson {
  mcpServers?: Record<string, unknown>;
  /** Ids of mcpServers entries YouCoded currently owns (see
   *  projectToClaudeJson). A plain top-level key, deliberately NOT nested
   *  inside individual server entries — see the OWNER DECISION comment on
   *  projectToClaudeJson below. */
  _youcodedOwnedMcpServers?: string[];
  [k: string]: unknown;
}

// Guard for the write path against a hand-edited/corrupted ~/.claude.json
// where `mcpServers` is present but not an object (e.g. a string or number).
// Spreading a non-object with `{ ...value }` does NOT throw in JS — for a
// string it produces numeric-indexed keys ('0', '1', ...) that would then get
// written back into the real config as garbage entries. Treat anything that
// isn't a plain object the same as "absent".
function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

type ManifestPlatform = 'macos' | 'windows' | 'linux';

function currentPlatform(): ManifestPlatform {
  if (process.platform === 'darwin') return 'macos';
  if (process.platform === 'win32') return 'windows';
  return 'linux';
}

// WHY: manifests in the wild use Node's names (darwin/win32) as well as the
// app's own (macos/windows). Normalise both so neither vocabulary is silently
// treated as "no match". Unknown words map to themselves and simply never match.
function normalizePlatform(p: string): string {
  const v = p.trim().toLowerCase();
  if (v === 'darwin' || v === 'mac' || v === 'osx') return 'macos';
  if (v === 'win32' || v === 'win') return 'windows';
  return v;
}

export function platformMatches(
  entry: Pick<McpManifestEntry, 'platform' | 'platforms'>,
  current: ManifestPlatform = currentPlatform(),
): boolean {
  // WHY: `platforms` (a list) wins when present because it is the more
  // specific declaration; an empty or non-array value falls back to the
  // single field so a malformed list never hides a server that also declares
  // `platform`.
  if (Array.isArray(entry.platforms) && entry.platforms.length > 0) {
    const list = entry.platforms.filter((p): p is string => typeof p === 'string').map(normalizePlatform);
    return list.includes('all') || list.includes(current);
  }
  const declared = entry.platform;
  if (!declared || declared === 'all') return true;
  return normalizePlatform(declared) === current;
}

// WHY: `${PACKAGE_DIR}` is the token the marketplace publisher skill and two
// published manifests use; only `{{plugin_root}}` used to be expanded, so those
// servers were written to ~/.claude.json with a literal placeholder and never
// started. Treat both as the plugin's install directory — this fixes the whole
// class (every community manifest that followed the publisher's guidance),
// not just the two known ones.
const PLUGIN_ROOT_TOKENS = /\{\{plugin_root\}\}|\$\{PACKAGE_DIR\}/g;
// Built by concatenation so it cannot be mistaken for a template string.
const UNEXPANDED_PACKAGE_DIR = '$' + '{PACKAGE_DIR}';

export function expandTokens(s: string, pluginRoot: string): string {
  // A replacer function, not a string, so a `$` in a real path is never read
  // as a replacement pattern.
  return s.replace(PLUGIN_ROOT_TOKENS, () => pluginRoot);
}

function readManifest(pluginDir: string): { entries: McpManifestEntry[]; pluginRoot: string } | null {
  const p = path.join(pluginDir, 'mcp-manifest.json');
  if (!fs.existsSync(p)) return null;
  try {
    const data = JSON.parse(fs.readFileSync(p, 'utf8'));
    const entries: McpManifestEntry[] = Array.isArray(data) ? data : (data.servers ?? []);
    return { entries, pluginRoot: pluginDir };
  } catch {
    return null;
  }
}

function listManifests(): Array<{ entries: McpManifestEntry[]; pluginRoot: string }> {
  // See claude-code-registry.listInstalledPluginDirs() — walks top-level
  // toolkit clone AND the marketplace subtree so both install paths surface.
  const out: Array<{ entries: McpManifestEntry[]; pluginRoot: string }> = [];
  for (const pluginDir of listInstalledPluginDirs()) {
    const m = readManifest(pluginDir);
    if (m) out.push(m);
  }
  return out;
}

// Fix (Finding 2, 2026-07-31): "absent" and "present but unreadable" are NOT
// the same thing, and the old catch-all `catch { return {}; }` treated them
// identically. `~/.claude.json` is Destin's LIVE Claude Code config — 59
// top-level keys, project history, onboarding state — and the caller below
// (reconcileMcp) writes `projected` back out whenever `mcpServers`/ownership
// differ from what it read. With a non-empty registry, an unreadable file
// (corrupt JSON from a partial write, EACCES, a read racing an external
// writer) reading as `{}` makes EVERY key look "changed" — the file gets
// atomically REPLACED with a bare `{mcpServers, _youcodedOwnedMcpServers}`
// skeleton, irrecoverably losing everything else, on every launch and every
// plugin install. Returning `null` here (never a throw — this must not crash
// launch) signals "abort, don't write" to the caller instead. Takes an
// explicit path (rather than reading the module-level CLAUDE_JSON constant)
// so tests can exercise this exact absent-vs-unreadable distinction against a
// temp file — never the real file.
export function readClaudeJsonFrom(filePath: string): ClaudeJson | null {
  if (!fs.existsSync(filePath)) return {};
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch (err) {
    log('ERROR', 'mcp-reconciler', 'unable to read an existing ~/.claude.json — aborting MCP reconcile rather than risk overwriting it', {
      path: filePath, error: String(err),
    });
    return null;
  }
}

function readClaudeJson(): ClaudeJson | null {
  return readClaudeJsonFrom(CLAUDE_JSON);
}

function writeClaudeJsonAtomic(data: ClaudeJson): void {
  const tmp = `${CLAUDE_JSON}.${process.pid}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(data, null, 2), 'utf8');
  fs.renameSync(tmp, CLAUDE_JSON);
}

/** Convert a manifest entry to the shape Claude Code expects in .claude.json. */
function buildServerConfig(
  entry: McpManifestEntry,
  pluginRoot: string,
  isWindows: boolean = process.platform === 'win32',
  expand: (s: string, root: string) => string = expandTokens,
): Record<string, unknown> | null {
  if (entry.type === 'http') {
    if (!entry.url) return null;
    return { type: 'http', url: entry.url };
  }
  // stdio (default)
  const rawCommand = isWindows && entry.command_windows
    ? entry.command_windows
    : entry.command;
  if (!rawCommand) return null;
  const config: Record<string, unknown> = {
    type: 'stdio',
    command: expand(rawCommand, pluginRoot),
  };
  if (entry.args) config.args = entry.args.map(a => expand(a, pluginRoot));
  if (entry.env) config.env = entry.env;
  return config;
}

/** What older builds wrote for this entry: `{{plugin_root}}` expanded but
 *  `${PACKAGE_DIR}` left literal. */
function legacyExpand(s: string, root: string): string {
  return s.replace(/\{\{plugin_root\}\}/g, () => root);
}

/**
 * Pure manifest-scan step of reconcileMcp (exported for tests). Mutates
 * `servers` in place and returns the counts.
 *
 * Additive-only, as before, with ONE narrow exception: an existing entry that
 * is exactly what an older build wrote for this same manifest entry AND still
 * carries the literal `${PACKAGE_DIR}` placeholder is replaced. WHY: that
 * entry can never have worked (nothing expands the token), and without the
 * repair every user who installed such a plugin before this fix would keep the
 * broken entry forever because the scan never overwrites. The deep-equality
 * check means an entry the user touched in any way is still left alone.
 */
export function applyManifestEntries(
  servers: Record<string, unknown>,
  manifests: Array<{ entries: McpManifestEntry[]; pluginRoot: string }>,
  opts: { platform?: ManifestPlatform; isWindows?: boolean } = {},
): { added: number; repaired: number; skippedPlatform: number; skippedManual: number; changed: boolean } {
  const platform = opts.platform ?? currentPlatform();
  const isWindows = opts.isWindows ?? process.platform === 'win32';
  let added = 0;
  let repaired = 0;
  let skippedPlatform = 0;
  let skippedManual = 0;
  for (const { entries, pluginRoot } of manifests) {
    for (const entry of entries) {
      if (!entry || !entry.name) continue;
      if (!platformMatches(entry, platform)) { skippedPlatform++; continue; }
      if (!entry.auto) { skippedManual++; continue; }
      const config = buildServerConfig(entry, pluginRoot, isWindows);
      if (!config) continue;
      const existing = servers[entry.name];
      if (existing) {
        // Never overwrite a user-configured entry — trust their customizations
        // (see the repair exception in this function's header).
        const legacy = buildServerConfig(entry, pluginRoot, isWindows, legacyExpand);
        const legacyText = JSON.stringify(legacy);
        if (legacyText.includes(UNEXPANDED_PACKAGE_DIR) && JSON.stringify(existing) === legacyText) {
          servers[entry.name] = config;
          repaired++;
        }
        continue;
      }
      servers[entry.name] = config;
      added++;
    }
  }
  return { added, repaired, skippedPlatform, skippedManual, changed: added + repaired > 0 };
}

/** Convert one RESOLVED registry server (secrets already decrypted) into the
 *  shape Claude Code expects in ~/.claude.json's mcpServers section. Distinct
 *  from buildServerConfig above: that one reads a plugin MANIFEST entry, this
 *  one reads a registry entry — different source shapes, same destination. */
function buildRegistryServerConfig(server: ResolvedMcpServer): Record<string, unknown> {
  if (server.transport.type === 'http') {
    const config: Record<string, unknown> = { type: 'http', url: server.transport.url };
    if (server.headers && Object.keys(server.headers).length > 0) config.headers = server.headers;
    return config;
  }
  // stdio
  const config: Record<string, unknown> = { type: 'stdio', command: server.transport.command };
  if (server.transport.args) config.args = server.transport.args;
  if (server.transport.cwd) config.cwd = server.transport.cwd;
  if (server.env && Object.keys(server.env).length > 0) config.env = server.env;
  return config;
}

/**
 * Pure projection of the YouCoded MCP registry into Claude Code's
 * ~/.claude.json shape. Kept pure so every case is unit-testable without ever
 * touching a real config file — `reconcileMcp` below is the only impure
 * caller (reads the file, calls this, writes atomically).
 *
 * OWNERSHIP (2026-07-30, spec 2026-07-30-native-mcp-design §3.3 — owner
 * decision overriding the original plan's per-entry marker):
 * YouCoded manages exactly the entries it marked as its own, tracked in a
 * TOP-LEVEL `_youcodedOwnedMcpServers: string[]` key, NOT a per-entry flag.
 * Claude Code demonstrably tolerates arbitrary top-level keys (Destin's real
 * file carries 59) but whether it tolerates an unknown key INSIDE an
 * mcpServers entry is unverified — writing one there could silently break MCP
 * loading in his live sessions if that schema turns out to be strict, so
 * server entries stay schema-clean.
 *
 * An entry it does not own is never modified or removed. An owned entry that
 * has left the registry (disabled or deleted) IS removed — that is how
 * disabling a server in YouCoded turns it off for Claude Code too, which is
 * what "YouCoded owns it" has to mean. The owned-id list is rebuilt fresh on
 * every call from THIS run's registry, so it can never drift from reality.
 *
 * COLLISION GUARD (fix pass, 2026-07-31): "an entry it does not own is never
 * modified" has to hold on the WRITE side too, not just the prune side above.
 * `sanitizeServerId()` applies no namespace prefix, so a human-picked id like
 * `gmail` from `claude mcp add gmail ...` collides exactly with a YouCoded
 * registry entry labeled "Gmail". Before writing a registry server's config,
 * check whether its id already exists in `mcpServers` AND is absent from
 * `previouslyOwned` — that combination means "something else put this key
 * here" (a hand-written entry, or a plugin-manifest-scanned one from the loop
 * in `reconcileMcp` that runs before this function). That id is skipped, not
 * overwritten, and reported back in `skippedCollisions` so a caller can offer
 * it as an explicit "adopt" action later (phase 2) instead of silently taking
 * it. An id already in `previouslyOwned` is a legitimate update, not a
 * collision, and proceeds as before.
 */
export function projectToClaudeJson(
  claudeJson: ClaudeJson,
  servers: ResolvedMcpServer[]
): { claudeJson: ClaudeJson; skippedCollisions: string[] } {
  const previouslyOwned = Array.isArray(claudeJson._youcodedOwnedMcpServers)
    ? claudeJson._youcodedOwnedMcpServers
    : [];
  // Non-object mcpServers (hand-edited/corrupted file: a string, a number)
  // is treated as absent rather than spread — see isPlainObject's comment.
  const mcpServers: Record<string, unknown> = {
    ...(isPlainObject(claudeJson.mcpServers) ? claudeJson.mcpServers : {}),
  };
  const stillInRegistry = new Set(servers.map((s) => s.id));

  // An id we owned last run that isn't in THIS run's registry has left
  // (disabled or deleted in YouCoded) — remove it. Deleting a key that's
  // already gone (the user hand-deleted the entry too) is a no-op, not a
  // crash — Set/delete on a missing key is always safe.
  for (const id of previouslyOwned) {
    if (!stillInRegistry.has(id)) delete mcpServers[id];
  }

  const nowOwned: string[] = [];
  const skippedCollisions: string[] = [];
  for (const server of servers) {
    // Defensive: projectToClaudeJson's contract is "enabled, resolved
    // servers only" regardless of what the caller passes — resolveAllEnabled()
    // already filters disabled ones, but a disabled entry must never be
    // projected even if a future caller passes list() by mistake.
    if (server.enabled === false) continue;
    if (server.credentialError) {
      // WHY: keep a previously working projection and its ownership while the
      // wallet is inaccessible; never publish partial credentials or adopt an
      // unowned entry. Independent servers still reconcile normally.
      if (previouslyOwned.includes(server.id) && Object.prototype.hasOwnProperty.call(mcpServers, server.id)) nowOwned.push(server.id);
      continue;
    }
    // A server synced from another device without its matching secret is
    // "needs setup" (McpRegistry.resolveEntry's missingSecrets) — projecting
    // it would hand Claude Code a command/header missing a required value,
    // so it's skipped until the secret actually resolves.
    if (server.missingSecrets && server.missingSecrets.length > 0) continue;
    // Collision guard — see this function's header comment. `mcpServers`
    // here still reflects the incoming file (pruning above only removes ids
    // NOT in `servers`, so it can never remove the very id being checked).
    const idAlreadyExists = Object.prototype.hasOwnProperty.call(mcpServers, server.id);
    if (idAlreadyExists && !previouslyOwned.includes(server.id)) {
      skippedCollisions.push(server.id);
      continue;
    }
    mcpServers[server.id] = buildRegistryServerConfig(server);
    nowOwned.push(server.id);
  }

  return {
    claudeJson: {
      ...claudeJson,
      mcpServers,
      _youcodedOwnedMcpServers: nowOwned,
    },
    skippedCollisions,
  };
}

export interface ReconcileMcpResult {
  added: number;
  skippedPlatform: number;
  skippedManual: number;
  manifestCount: number;
  /** Registry server ids skipped this run because the id collided with an
   *  entry (hand-written by the user, or manifest-scanned) that YouCoded does
   *  not own — see projectToClaudeJson's collision guard. Empty in the common
   *  case; a non-empty list is worth logging so a collision doesn't go
   *  unnoticed even though phase 1 has no UI to surface it yet. */
  skippedCollisions: string[];
  /** Fix (Finding 2): true when ~/.claude.json existed but could not be read
   *  (corrupt JSON, EACCES, a partial read racing an external writer) — the
   *  reconcile aborted WITHOUT writing anything, real cause already logged by
   *  readClaudeJsonFrom. Every other field is the zero-work default in this
   *  case (manifests are scanned above the read, so manifestCount can still
   *  be non-zero even though nothing was written). */
  aborted: boolean;
}

export async function reconcileMcp(): Promise<ReconcileMcpResult> {
  const manifests = listManifests();
  const claudeJson = readClaudeJson();
  if (claudeJson === null) {
    // Fix (Finding 2): abort rather than proceed from `{}` — see
    // readClaudeJsonFrom's comment. Returning a result (never throwing) keeps
    // this from blocking the rest of app startup / a plugin install, which
    // both call reconcileMcp() and only log its outcome.
    return { added: 0, skippedPlatform: 0, skippedManual: 0, manifestCount: manifests.length, skippedCollisions: [], aborted: true };
  }
  const servers = (claudeJson.mcpServers as Record<string, unknown>) || {};

  // Legacy path (decomposition v3 §9.3): plugin-bundled servers declared in
  // mcp-manifest.json. Additive-only — never removed, and never overwritten
  // except the one never-worked `${PACKAGE_DIR}` repair documented on
  // applyManifestEntries. Folding this into the registry (one ownership rule
  // for both sources) was considered and deliberately deferred — see
  // task-7-report.md.
  const { added, skippedPlatform, skippedManual, changed: manifestChanged } =
    applyManifestEntries(servers, manifests);
  claudeJson.mcpServers = servers;

  // New path (Task 7): project the YouCoded-owned MCP registry on top. A
  // fresh, cheap NativeHome/SecretsStore pair — same construction ipc-
  // handlers.ts already uses for ProviderRegistry; no shared connections to
  // pool here (that's McpManager's job for LIVE sessions), just a read.
  const registry = new McpRegistry(new NativeHome(), new SecretsStore(app.getPath('userData')));
  const resolved = await registry.resolveAllEnabled();
  const { claudeJson: projected, skippedCollisions } = projectToClaudeJson(claudeJson, resolved);

  // added/skippedPlatform/skippedManual intentionally keep their ORIGINAL
  // meaning (manifest-scan counts only) rather than being redefined to
  // include registry churn — the registry re-projects its owned entries on
  // every call by design, so counting that as "added" every launch would be
  // misleading telemetry, not a bug fix.
  const mcpServersChanged = JSON.stringify(claudeJson.mcpServers) !== JSON.stringify(projected.mcpServers);
  const ownershipChanged = JSON.stringify(claudeJson._youcodedOwnedMcpServers ?? []) !== JSON.stringify(projected._youcodedOwnedMcpServers ?? []);

  if (manifestChanged || mcpServersChanged || ownershipChanged) {
    writeClaudeJsonAtomic(projected);
  }
  return { added, skippedPlatform, skippedManual, manifestCount: manifests.length, skippedCollisions, aborted: false };
}
